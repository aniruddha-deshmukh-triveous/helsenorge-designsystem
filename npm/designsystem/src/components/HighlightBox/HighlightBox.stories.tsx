import React from 'react';

import { ComponentStory, ComponentMeta } from '@storybook/react';

import HighlightBox, { HighlightBoxSize } from './HighlightBox';
import PdfFile from '../Icons/PdfFile';

export default {
  title: 'HighlightBox',
  component: HighlightBox,
  argTypes: {
    children: {
      control: 'text',
      defaultValue:
        'Risiko for alvorleg sjukdom aukar med alderen og underliggande sjukdomar, og menn har høgare risiko enn kvinner. Unge personar utan kjende risikofaktorar kan også få eit alvorleg forløp av sjukdomen. Det er meir sannsynleg at dette skjer dersom dei har underliggande sjukdomar.',
    },
    color: {
      control: 'select',
      options: ['white', 'blueberry', 'cherry', 'neutral', 'kiwi', 'plum'],
      defaultValue: 'white',
    },
    size: {
      control: 'select',
      options: HighlightBoxSize,
      defaultValue: HighlightBoxSize.medium,
    },
    htmlMarkup: {
      control: 'select',
      options: ['div', 'nav', 'section', 'article', 'span'],
      defaultValue: 'div',
    },
  },
} as ComponentMeta<typeof HighlightBox>;

export const Default: ComponentStory<typeof HighlightBox> = (args: any) => (
  <div>
    <HighlightBox {...args}>{args.children}</HighlightBox>
  </div>
);

export const WithIcon: ComponentStory<typeof HighlightBox> = (args: any) => (
  <div>
    <HighlightBox {...args} svgIcon={PdfFile}>
      {args.children}
    </HighlightBox>
  </div>
);
