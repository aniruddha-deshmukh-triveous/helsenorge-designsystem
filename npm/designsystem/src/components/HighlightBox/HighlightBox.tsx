import React from 'react';
import classNames from 'classnames';
import { PaletteNames } from '../../theme/palette';

import styles from './styles.module.scss';
import { AnalyticsId } from '../../constants';

export type HighlightBoxColors = Extract<PaletteNames, 'white' | 'blueberry' | 'cherry' | 'neutral' | 'kiwi' | 'plum'>;

export enum HighlightBoxSize {
  medium = 'medium',
  large = 'large',
  fluid = 'fluid',
}

export type HighlightBoxTags = Exclude<
  keyof HTMLElementTagNameMap,
  'dir' | 'font' | 'frame' | 'frameset' | 'marquee' | 'applet' | 'basefont'
>;

interface HighlightBoxProps {
  /** What's in the box? */
  children: React.ReactNode;
  /** Changes the background color. Default: white */
  color?: HighlightBoxColors;
  /** Changes the size. Default: medium */
  size?: keyof typeof HighlightBoxSize;
  /** Changes the underlying element. Default: div */
  htmlMarkup?: HighlightBoxTags;
  /** Sets the data-testid attribute. */
  testId?: string;
}

const HighlightBox: React.FC<HighlightBoxProps> = props => {
  const { children, color = 'white', size = HighlightBoxSize.medium, testId, htmlMarkup = 'div' } = props;

  const containerClassName = classNames(styles.highlightbox, styles[`highlightbox--${color}`], styles[`highlightbox--${size}`]);

  const CustomTag = htmlMarkup;

  if (props.size === HighlightBoxSize.fluid) {
    return (
      <CustomTag className={containerClassName} data-testid={testId}>
        {children}
      </CustomTag>
    );
  }

  return (
    <div className={containerClassName}>
      <div className={styles.highlightbox__row}>
        <CustomTag className={styles.highlightbox__col} data-testid={testId} data-analyticsid={AnalyticsId.HighlightBox}>
          {children}
        </CustomTag>
      </div>
    </div>
  );
};

export default HighlightBox;
