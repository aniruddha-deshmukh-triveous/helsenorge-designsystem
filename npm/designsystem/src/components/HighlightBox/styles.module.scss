@import '../../scss/_spacers.scss';
@import '../../scss/_breakpoints.scss';
@import '../../scss/_palette.scss';
@import '../../scss/_grid.scss';

.highlightbox {
  $box: &;
  $colors: 'blueberry', 'cherry', 'neutral', 'kiwi', 'plum';

  // Medium og large har bakgrunnsfarge på den indre wrapperen
  // slik at bakgrunnsfargen ikke går helt ut i kantene
  &--medium,
  &--large {
    @include make-container();
    @include make-container-max-widths();
    @each $color in $colors {
      &#{$box}--#{$color} {
        #{$box}__content-wrapper {
          background-color: map-get($palette-map, #{$color}50);
        }
      }
    }

    // Hvit har sin egen regel siden den ikke har valører
    &#{$box}--white {
      #{$box}__content-wrapper {
        background-color: $white;
      }
    }
  }

  // large skal også ha bakgrunnsfarge på den ytre containeren på de minste skjermnene
  // slik at bakgrunnsfargen går helt ut til skjermkanten
  &--large {
    @each $color in $colors {
      &#{$box}--#{$color} {
        @include media-breakpoint-down(xs) {
          background-color: map-get($palette-map, #{$color}50);
        }
      }
    }

    &#{$box}--white {
      @include media-breakpoint-down(xs) {
        background-color: $white;
      }
    }
  }

  &--medium {
    #{$box}__content-wrapper {
      padding: getSpacer(m) getSpacer(s) getSpacer(l) getSpacer(s);
      @media (min-width: map-get($grid-breakpoints, md)) {
        padding: getSpacer(m) getSpacer(m) getSpacer(l) getSpacer(m);
      }
      @media (min-width: map-get($grid-breakpoints, lg)) {
        padding: getSpacer(l);
      }
    }
  }

  &--large {
    #{$box}__content-wrapper {
      padding-top: getSpacer(m);
      padding-bottom: getSpacer(l);
      @media (min-width: map-get($grid-breakpoints, lg)) {
        padding-top: getSpacer(l);
        padding-bottom: getSpacer(xl);
      }
    }
  }

  // fluid skal ha bakgrunnsfarge på den ytre containeren
  &--fluid {
    padding: getSpacer(m) getSpacer(s) getSpacer(l) getSpacer(s);

    @media (min-width: map-get($grid-breakpoints, lg)) {
      padding: getSpacer(m) getSpacer(m) getSpacer(l) getSpacer(m);
    }

    @each $color in $colors {
      &#{$box}--#{$color} {
        background-color: map-get($palette-map, #{$color}50);
      }
    }
    &#{$box}--white {
      background-color: $white;
    }
  }

  &--has-icon#{&}--medium {
    #{$box}__content-wrapper {
      display: flex;
      @media (min-width: map-get($grid-breakpoints, lg)) {
        padding-left: getSpacer(m);
      }
    }
  }

  &--has-icon#{&}--large {
    #{$box}__col {
      display: flex;
    }
    #{$box}__content-wrapper {
      @media (min-width: map-get($grid-breakpoints, sm)) {
        padding-left: getSpacer(s);
      }
      @media (min-width: map-get($grid-breakpoints, md)) {
        padding-left: getSpacer(m);
      }
      @media (min-width: map-get($grid-breakpoints, lg)) {
        padding-left: 0;
        padding-top: getSpacer(l);
        padding-bottom: getSpacer(xl);
      }
    }

    #{$box}__icon {
      @media (min-width: map-get($grid-breakpoints, lg)) {
        @include make-col-ready();
        @include make-col(2);
        display: flex;
        justify-content: end;
        margin-right: 0;
        padding-right: getSpacer(s);
      }
    }

    #{$box}__content {
      @media (min-width: map-get($grid-breakpoints, lg)) {
        @include make-col-ready();
        @include make-col(8);
      }
    }
  }

  &--has-icon#{&}--fluid {
    display: flex;
  }

  &__icon {
    margin-right: getSpacer(s);
  }

  &__row {
    @include make-row();
  }

  &__col {
    @include make-col-ready();
    @include make-col(12);

    &--offset {
      @media (min-width: map-get($grid-breakpoints, sm)) {
        @include make-col(10);
        @include make-col-offset(1);
      }
      @media (min-width: map-get($grid-breakpoints, lg)) {
        @include make-col(8);
        @include make-col-offset(2);
      }
    }

    &--large-with-icon {
      @media (min-width: map-get($grid-breakpoints, sm)) {
        @include make-col(11);
        @include make-col-offset(0);
      }

      @media (min-width: map-get($grid-breakpoints, lg)) {
        @include make-col(12);
      }
    }
  }
}
