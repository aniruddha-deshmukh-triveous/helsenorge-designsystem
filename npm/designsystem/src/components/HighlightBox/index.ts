import HighlightBox from './HighlightBox';
export * from './HighlightBox';
export default HighlightBox;
