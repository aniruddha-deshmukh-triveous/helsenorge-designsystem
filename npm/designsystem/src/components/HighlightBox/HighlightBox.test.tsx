import React from 'react';

import { render, screen } from '@testing-library/react';

import HighlightBox from './HighlightBox';
import PdfFile from '../Icons/PdfFile';

describe('Gitt at HighlightBox skal rendres', (): void => {
  describe('Når HighlightBox skal vises vanlig', (): void => {
    test('Så rendres HighlightBox riktig', (): void => {
      render(<HighlightBox testId="boksen">Teksten sin</HighlightBox>);

      const box = screen.getByTestId('boksen');
      expect(box).toBeVisible();

      const text = screen.getByText('Teksten sin');
      expect(text).toBeVisible();
    });
  });
  describe('Når HighlightBox skal vises som medium', (): void => {
    test('Så rendres HighlightBox riktig', (): void => {
      const { container } = render(
        <HighlightBox color="blueberry" size="medium">
          <p>Avsnitt</p>
        </HighlightBox>
      );

      expect(container).toMatchSnapshot();
    });
  });
  describe('Når HighlightBox skal vises som large', (): void => {
    test('Så rendres HighlightBox riktig', (): void => {
      const { container } = render(
        <HighlightBox color="blueberry" size="large">
          <p>Avsnitt</p>
        </HighlightBox>
      );

      expect(container).toMatchSnapshot();
    });
  });
  describe('Når HighlightBox skal vises som fluid', (): void => {
    test('Så rendres HighlightBox riktig', (): void => {
      const { container } = render(
        <HighlightBox color="blueberry" size="fluid">
          <p>Avsnitt</p>
        </HighlightBox>
      );

      expect(container).toMatchSnapshot();
    });
  });
  describe('Når HighlightBox skal vises som et nav-element', (): void => {
    test('Så har komponenten riktig rolle', (): void => {
      render(
        <HighlightBox testId="boksen" htmlMarkup="nav">
          Teksten sin
        </HighlightBox>
      );

      const box = screen.getByRole('navigation');
      expect(box).toBeVisible();
    });
  });
  describe('Når HighlightBox skal vises med HTML som children', (): void => {
    test('Så er children rendret riktig', (): void => {
      render(
        <HighlightBox testId="boksen" htmlMarkup="nav">
          <h1>Jeg er en tittel</h1>
        </HighlightBox>
      );

      const title = screen.getByRole('heading', { name: 'Jeg er en tittel' });
      expect(title).toBeVisible();
    });
  });

  describe('Når HighlightBox skal vises med ikon', (): void => {
    test('Så er ikonet med', (): void => {
      render(
        <HighlightBox testId="boksen" svgIcon={PdfFile}>
          <h1>Jeg er en tittel</h1>
        </HighlightBox>
      );

      const icon = screen.getByRole('presentation', { hidden: true });
      expect(icon).toBeInTheDocument();
    });
  });
  describe('Når HighlightBox skal vises med custom className', (): void => {
    test('Så er className med', (): void => {
      render(
        <HighlightBox className="test-av-custom-classname">
          <h1>Jeg er en tittel</h1>
        </HighlightBox>
      );

      const box = screen.getByTestId('highlightbox-wrapper');
      expect(box).toHaveClass('test-av-custom-classname');
    });
  });
  describe('Når HighlightBox skal vises som fluid med custom className', (): void => {
    test('Så er className med', (): void => {
      render(
        <HighlightBox className="test-av-custom-classname" size={'fluid'} testId="fluid-highlightbox">
          <h1>Jeg er en tittel</h1>
        </HighlightBox>
      );

      const box = screen.getByTestId('fluid-highlightbox');
      expect(box).toHaveClass('test-av-custom-classname');
    });
  });
});
