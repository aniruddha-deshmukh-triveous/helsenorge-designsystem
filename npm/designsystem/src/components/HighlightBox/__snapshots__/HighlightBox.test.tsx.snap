// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Gitt at HighlightBox skal rendres Når HighlightBox skal vises som fluid Så rendres HighlightBox riktig 1`] = `
<div>
  <div
    class="highlightbox highlightbox--blueberry highlightbox--fluid"
  >
    <p>
      Avsnitt
    </p>
  </div>
</div>
`;

exports[`Gitt at HighlightBox skal rendres Når HighlightBox skal vises som large Så rendres HighlightBox riktig 1`] = `
<div>
  <div
    class="highlightbox highlightbox--blueberry highlightbox--large"
  >
    <div
      class="highlightbox__row"
    >
      <div
        class="highlightbox__col"
        data-analyticsid="highlight-box"
      >
        <p>
          Avsnitt
        </p>
      </div>
    </div>
  </div>
</div>
`;

exports[`Gitt at HighlightBox skal rendres Når HighlightBox skal vises som medium Så rendres HighlightBox riktig 1`] = `
<div>
  <div
    class="highlightbox highlightbox--blueberry highlightbox--medium"
  >
    <div
      class="highlightbox__row"
    >
      <div
        class="highlightbox__col"
        data-analyticsid="highlight-box"
      >
        <p>
          Avsnitt
        </p>
      </div>
    </div>
  </div>
</div>
`;
